be/src/storage/compaction_scheduler.cpp

```diff
@@ -0,0 +1,242 @@
+    if (config::max_compaction_task_per_disk >= 0 && num >= config::max_compaction_task_per_disk) {
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id()
+                  << " for limit of compaction task per disk. disk path:" << data_dir->path()
+                  << ", running num:" << num;
+        return false;
+    }
+    // found a qualified tablet
+    // qualified tablet will be removed from candidates
+    need_reset_task = false;
+    return true;
+}
+
+bool CompactionScheduler::_check_precondition(Tablet* tablet) {
+    if (!tablet) {
+        LOG(WARNING) << "null tablet";
+        return false;
+    }
+    if (!tablet->need_compaction()) {
+        // check need compaction
+        // if it is false, skip this tablet and remove it from candidate
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " because need_compaction is false";
+        return false;
+    }
+
+    if (tablet->tablet_state() != TABLET_RUNNING) {
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " because tablet state is not RUNNING";
+        return false;
+    }
+
+    // check for alter task for safety
+    // maybe this logic can be removed if tablet is the schema change dest tablet, need_compaction will never return true
+    AlterTabletTaskSharedPtr cur_alter_task = tablet->alter_task();
+    if (cur_alter_task != nullptr && cur_alter_task->alter_state() != ALTER_FINISHED &&
+        cur_alter_task->alter_state() != ALTER_FAILED) {
+        TabletManager* tablet_manager = StorageEngine::instance()->tablet_manager();
+        TabletSharedPtr related_tablet = tablet_manager->get_tablet(cur_alter_task->related_tablet_id());
+        if (related_tablet != nullptr && tablet->creation_time() > related_tablet->creation_time()) {
+            // Current tablet is newly created during schema-change or rollup, skip it
+            LOG(INFO) << "skip tablet:" << tablet->tablet_id() << " because it is newly-created for schema change";
+            return false;
+        }
+    }
+
+    std::shared_ptr<CompactionTask> compaction_task = tablet->get_compaction(false);
+    if (compaction_task) {
+        // tablet already has a running compaction task, skip it
+        LOG(INFO) << "skip tablet:" << tablet->tablet_id()
+                  << " because there is another running compaction task:" << compaction_task->task_id();
+        return false;
+    }
+    return true;
+}
+
+bool CompactionScheduler::_can_do_compaction(Tablet* tablet, bool* need_reschedule) {
+    // when the following conditions fail, should not reschedule the candidate tablet
+    *need_reschedule = false;
+    bool precondition_ok = _check_precondition(tablet);
+    if (!precondition_ok) {
+        return false;
+    }
+
+    // when the following conditions fail, should reschedule the candidate tablet
+    *need_reschedule = true;
+    DataDir* data_dir = tablet->data_dir();
+    if (data_dir->reach_capacity_limit(0)) {
```

Comment:
I think we should reschedule it because the data dir's capacity may go down. Maybe here the limit should be change to another config. It seems that 0 is not a reasonable config because compaction will cost disk capacity.